src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs

```diff
@@ -1131,8 +1134,21 @@ private BinaryOperatorAnalysisResult BinaryOperatorOverloadResolution(BinaryOper
             return possiblyBest;
         }
 
-        private bool IsDefaultLiteralAllowedInBinaryOperator(BinaryOperatorKind kind, BoundExpression left, BoundExpression right)
+        private bool IsTypelessExpressionAllowedInBinaryOperator(BinaryOperatorKind kind, BoundExpression left, BoundExpression right)
         {
+            // The default literal is only allowed with equality operators
+            // and both operands cannot be typeless at the same time.
+
+            if (left.IsTypelessNew())
+            {
+                return !right.IsTypelessNew() && !right.IsLiteralDefault();
+            }
+
+            if (right.IsTypelessNew())
+            {
+                return !left.IsLiteralDefault();
+            }
+
```

Comment:
Please add a PROTOTYPE comment that this code is only necessary because typeless new and default expressions are improperly represented as having a type in the bound node.

Reply:
typeless new is correctly defined with always-null type. I think we should add ConvertedDefaultExpression or something to fix this.